File: Sources/Overlays/_Testing_WinSDK/Attachments/GDI+.swift
//
// This source file is part of the Swift.org open source project
//
// Copyright (c) 2024 Apple Inc. and the Swift project authors
// Licensed under Apache License v2.0 with Runtime Library Exception
//
// See https://swift.org/LICENSE.txt for license information
// See https://swift.org/CONTRIBUTORS.txt for Swift project authors
//

#if os(Windows)
@_spi(Experimental) import Testing
import _Testing_WinSDK_GDIPlus

internal import WinSDK

/// A type describing errors that can be thrown by GDI+.
enum GDIPlusError: Error {
  /// A GDI+ status code.
  case status(SWTGDIPlusStatusCode)

  /// The testing library failed to create an in-memory stream.
  case streamCreationFailed(HRESULT)

  /// The testing library failed to get an in-memory stream's underlying buffer.
  case globalFromStreamFailed(HRESULT)
}

extension GDIPlusError: CustomStringConvertible {
  var description: String {
    switch self {
    case let .status(status):
      "Could not create the corresponding GDI+ image (Gdiplus.Status \(status.rawValue))."
    case let .streamCreationFailed(result):
      "Could not create an in-memory stream (HRESULT \(result))."
    case let .globalFromStreamFailed(result):
      "Could not access the buffer containing the encoded image (HRESULT \(result))."
    }
  }
}

/// Call the given GDI+ function and, if it failed, throw an error.
/// 
/// - Parameters:
///   - expression: An expression to evaluate that returns a GDI+ status code.
/// 
/// - Returns: Always returns `true`.
/// 
/// - Throws: An instance of ``GDIPlusError`` if `expression` failed.
@discardableResult
func call(_ expression: @autoclosure () -> SWTGDIPlusStatusCode) throws(GDIPlusError) -> Bool {
  switch expression() {
  case .ok:
    return true
  case let statusCode:
    throw GDIPlusError.status(statusCode)
  }
}

// MARK: -

/// Call a function while GDI+ is set up on the current thread.
/// 
/// - Parameters:
///   - body: The function to invoke.
///
/// - Returns: Whatever is returned by `body`.
///
/// - Throws: Whatever is thrown by `body`.
func withGDIPlus<R>(_ body: () throws -> R) throws -> R {
  // "Escape hatch" if the program being tested calls GdiplusStartup() itself in
  // some way that is incompatible with our assumptions about it.
  if Environment.flag(named: "SWT_GDIPLUS_STARTUP_ENABLED") == false {
    return try body()
  }

  var token = ULONG_PTR(0)  
  try call(swt_GdiplusStartup(&token))
  defer {
    swt_GdiplusShutdown(token)
  }

  return try body()
}
#endif

File: Sources/_TestingInternals/GDI+.cpp
//
// This source file is part of the Swift.org open source project
//
// Copyright (c) 2024 Apple Inc. and the Swift project authors
// Licensed under Apache License v2.0 with Runtime Library Exception
//
// See https://swift.org/LICENSE.txt for license information
// See https://swift.org/CONTRIBUTORS.txt for Swift project authors
//

#include "GDI+.h"

#include <gdiplus.h>
#include <shlwapi.h>

using Gdiplus;

ULONG_PTR swt_gdiplus_startup(int *outError) {
  ULONG_PTR result = nullptr;

  GdiplusStartupInput input;
  auto status = GdiplusStartup(&result, &input, nullptr);

  if (status != Ok) {
    *outError = static_cast<int>(status);
  }
  return result;
}

void swt_gdiplus_shutdown(ULONG_PTR *token) {
  (void)GdiplusShutdown(token);
}

void *swt_gdiplus_createImageFromHBITMAP(HBITMAP bitmap, HPALETTE palette) {
  return Bitmap::FromHBITMAP(bitmap, palette);
}

void swt_gdiplus_destroyImage(void *image) {
  auto bitmap = reinterpret_cast<Bitmap *>(image);
  delete bitmap;
}

void *swt_gdiplus_copyBytes(void *image, const CLSID *clsid, size_t *outByteCount, int *outError) {
  auto bitmap = reinterpret_cast<Bitmap *>(image);

  // Create an IStream in memory and save the image to it.
  auto stream = SHCreateMemStream(nullptr, 0);
  auto status = bitmap->Save(stream, clsid);

  // Read back from the stream into 
  (void)stream->Seek(0, STREAM_SEEK_SET, nullptr);

  stream->Release();

  if (status != Ok) {
    *outError = static_cast<int>(status);
  }
  return nullptr;
}
